common/wuanalysis/anarule.cpp

```diff
@@ -165,11 +165,40 @@ class IoSkewRule : public AActivityRule
     const char * category;
 };
 
+class keyedJoinExcessRejectedRowsRule : public ActivityKindRule
+{
+public:
+    keyedJoinExcessRejectedRowsRule() : ActivityKindRule(TAKkeyedjoin) {}
+
+    virtual bool check(PerformanceIssue & result, IWuActivity & activity, const WuAnalyseOptions & options) override
+    {
+        stat_type preFiltered = activity.getStatRaw(StNumPreFiltered);
+        if (preFiltered)
+        {
+            IWuEdge * inputEdge = activity.queryInput(0);
+            stat_type rowscnt = inputEdge->getStatRaw(StNumRowsProcessed);
+            stat_type preFilteredPer = (preFiltered*100/ rowscnt); // prefiltered %
+            if (  preFilteredPer > preFilteredThreshold)
+            {
+                IWuActivity * inputActivity = inputEdge->querySource();
+                stat_type timeAvgLocalExecute = inputActivity->getStatRaw(StTimeLocalExecute, StAvgX);
+                stat_type cost = timeAvgLocalExecute * preFilteredPer / 100;
+                result.set(ANA_KEYED_JOIN_EXCESSREJECTS_ID, cost, "Large number of rows from left dataset rejected in keyed join");
+                updateInformation(result, activity);
+                return true;
+            }
+        }
+        return false;
+    }
+private:
+    static const stat_type preFilteredThreshold = 20; //percent
```

Comment:
Is this a sensible default?  I think this means if >20% rows are rejected then the warning is reported.  For keyed join that may be fine, for index read it would too high.

This should probably be a member of WuAnalyseOptions.